Nova Driver Prototype for Google Cloud Platform

sanketsudake opened the pull request:
Attached pull request focuses on performing basic functionalities in GCE(Google Compute Engine) through nova driver like spawning, rebooting instances etc.

Basic instance operations like spawning, deleting, start, stop working with this driver. Currently most of glance and neutron specific things are hard-coded. Once Glance and Neutron support is in place, we should be able to perform pretty much matching behavior w.r.t. AWS.

Commit:
Nova driver prototype for Google Compute Engine

Commit:
Including GCE client python requirements. Renamed aws requirements

pacharya-pf9 commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,807 @@
+"""Connection to the Google Cloud Platform -  GCE service"""
+
+# import base64
```
Please remove the commented code

pacharya-pf9 commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,807 @@
+"""Connection to the Google Cloud Platform -  GCE service"""
```
Please add open source license header. See https://github.com/platform9/openstack-omni/blob/master/nova/ec2/ec2driver.py#L2

sanketsudake replied:
Added License in all files

pacharya-pf9 commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,807 @@
+                    "PROVISIONING", "STAGING", "STOPPING", "SUSPENDING",
+                    "TERMINATED"
+            ]:
+                continue
```
Add a debug log state here for the instance being skipped and its state.

sanketsudake replied:
Done

pacharya-pf9 commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,807 @@
+        # Update gce_id in openstack instance['metadata']
+        instance['metadata'].update({'gce_id': gce_instance['name']})
+        # TODO: Fix
+        instance['metadata'].update({'public_ip_address': '104.197.111.111'})
```
Why hardcode the IP address?

sanketsudake replied:
There is no direct GCE API to get public IP address from instance info. Working on it. Currently this is added as metadata only.

pacharya-pf9 commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,807 @@
+
+        """
+        compute, project, zone = self.gce_svc, self.gce_project, self.gce_zone
+        instance_name = instance['name']
```
Since Platform9 is upgrading OpenStack to Newton release it is better to . notation instead of [].
Please change the line to instance.name. This should also work with Liberty code.

sanketsudake replied:
Done

pacharya-pf9 commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,807 @@
+        :param instance: nova.objects.instance.Instance
+        :param image_id: Reference to a pre-created image holding the snapshot.
+        """
+        pass
```
Snapshot operation in OpenStack creates an image from the instance. Is this not supported by GCE?

sanketsudake replied:
Planning to have implementation with glance and cinder drivers in place.

pacharya-pf9 commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,807 @@
+
+    def suspend(self, context, instance):
+        """
+        Boto doesn't support suspend and cannot save system state and hence
```
Please update the documentation here:
boto->gce

sanketsudake replied:
Done

pacharya-pf9 commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,807 @@
+            'vcpus_used': 0,
+            'memory_mb_used': 0,
+            'local_gb_used': 0,
+            'hypervisor_type': 'EC2',
```
change hypervisor type

sanketsudake replied:
Fixed at all places required

pacharya-pf9 commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,807 @@
+        return self._uuid_to_gce_instance.keys()
+
+
+class GCEVirtAPI(virtapi.VirtAPI):
```
This class may not be required now. can you please check if driver works properly after removing this class?

sanketsudake replied:
Removed. Don't see any dependencies.

pacharya-pf9 reviewed (changes requested):
Hi Sanket,

I have taken an initial look at the code and added by comments. Will provide more feedback if needed.

Thanks,
Pushkar

Commit:
Add Licence to all driver files

Commit:
Remove unwanted imports in GCE driver

Commit:
GCE Nova driver: Fixes in  spawing instance

Commit:
GCE Nova driver: Use new notation for accessing instance fields

Commit:
GCE Nova driver: Adding logging insted print in gceutils

poojaghumre commented on nova/gce/constants.py:
```diff
@@ -0,0 +1,101 @@
+    "TERMINATED": power_state.CRASHED
+}
+
+GCE_FLAVOR_MAP = {
```
Can you add a comment/link here on where this list of flavors was derived from?

sanketsudake replied:
Added Get machines info API which generates required Flavor map. I have fetched this through API only. Refer `gceutils.get_machines_info` for more details.

poojaghumre commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,803 @@
+
+import gceutils
+
+from .constants import GCE_FLAVOR_MAP, GCE_STATE_MAP
```
nit: For consistency, can you make this a fully qualified import?

sanketsudake replied:
Done. I have added this relative import for temporary purposes.

poojaghumre commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,803 @@
+        # Update gce_id in openstack instance['metadata']
+        instance.metadata.update({'gce_id': gce_instance['name']})
+        # TODO: Fix
+        instance.metadata.update({'public_ip_address': '104.197.111.111'})
```
Is this public ip address not available from the GCE instance object?

sanketsudake replied:
Fixed. Getting public address from GCE now. This may changes after once neutron driver is in place.

poojaghumre commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,803 @@
+        instance.
+        :param instance: nova.objects.instance.Instance
+        """
+        self.power_off(instance)
```
Can you just add a log here suggesting power_off is being requested instead of suspend for clarity?

sanketsudake replied:
Added logging at different places required.

poojaghumre commented on nova/gce/driver.py:
```diff
@@ -0,0 +1,803 @@
+        instance.
+        :param instance: nova.objects.instance.Instance
+        """
+        self.power_on(context, instance, network_info, block_device_info)
```
Same comment as above for suspend

poojaghumre reviewed (commented):
Hi Sanket, providing some minor review comments below. Please take a look.
Thanks!

Commit:
GCE Nova Driver Changes
1. Get machines info using GCE API
2. Update public IP address in instance metadata
3. Remove Virt API class(not required)
4. Added missing logging inside few of driver functions

sanketsudake commented:
@pacharya-pf9  @poojaghumre  I have resolved most of your comments. Let me know if missed anything.
